.circleci/integration.sh

```diff
@@ -25,10 +31,23 @@ for version in $VERSIONS; do
     
     echo "--------------------  Testing with TEST_TLS=${TEST_TLS}"
 
-    python3 tests/test.py -v;
-    if [[ "$?" != "0" ]]; then
-        echo "golang-$version tls_on=${TEST_TLS}" >> "$FILEPATH_ERROR_FLAG";
-    else
-        echo "no errors";
-    fi
+    for iteration in {1..3}; do
+        context="${iteration}-golang-${version}-tls-${TEST_TLS}"
+        export TEST_XMLOUTPUT="${TEST_OUTPUT_FOLDER}/${context}.xml"
+        LOG_OUTPUT="${TEST_OUTPUT_FOLDER}/${context}.log"
+        timeout ${TEST_RUN_TIMEOUT} python3 tests/test.py -v | tee "${LOG_OUTPUT}";
+        if [[ "$?" != "0" ]]; then
+            status="$?"
```

Comment:
The status should be saved immediately after the failed command. The `$?` here contains the result of the test performed by `if`: i.e, always zero as we're in the success branch.

```diff
         timeout ${TEST_RUN_TIMEOUT} python3 tests/test.py -v | tee "${LOG_OUTPUT}";
-        if [[ "$?" != "0" ]]; then
-                status="$?"
+        status="$?"
+        if [[ "$status" != "0" ]]; then
```

Reply:
And we should also add

```bash
set -o pipefail
```

somewhere at the start of the script. Otherwise the pipes return the exit code of the last command: `tee` in this case. Even if `python` terminates unsuccessfully, `tee` will just see the closed input and will exit cleanly.

Reply:
thanks. nice catch

Reply:
If we're going to make this script really safe, I'd suggest to add `set -uo pipefail` at the start of the script.
It would add checks for unset variables (if that variables are there, they should have default values like a `${MYVAR:-defaultval}`). Often it is useful also to add `-e`, which terminate script on any command fail. In such case we have to handle expected errors inside `if` clauses.

Reply:
I don't want to force use default values in bash scripts because most of them use environment variables from circleci's config `.circleci/config.yaml`. with such checks, we should to duplicate the same values from `config.yaml` to each bash script